cleanup, formatting and styling of volumes concept (#24077)
kubernetes/website: content/en/docs/concepts/storage/volumes.md

<!-- overview -->

On-disk files in a container are ephemeral, which presents some problems for
non-trivial applications when running in containers. First, if a container

> crashes, the files in the container are lost. The kubelet restarts the container

Author: Hi. This has been a good (yes tedious to edit and review) exercise. The page needs basic updating as
there are many minor formatting nits. The page is long (+1400 lines). I thought it is a good place
to look at the styling and usage of terms (resources, drivers, plugins, ...). I realized that is is hard to
define one rule that fits all. I chose to "try out" a few styling options to test the content.

The question about placing the kubelet in code font has come up before. If I remember correctly, it
was decided that the kubelet (refer to as "the kubelet" in text) should not be formatted as code.
If a sentence begins with a tool name such as `kubectl`, you do not have to capitalize this name.
Try and reword the sentence so that the sentence does not begin with this name.

<!-- overview -->

On-disk files in a container are ephemeral, which presents some problems for
non-trivial applications when running in containers. One problem
is the loss of files when a container crashes. The kubelet restarts the container
but with a clean state. A second problem occurs when sharing files
between containers running together in a pod.
The Kubernetes {{< glossary_tooltip text="volume" term_id="volume" >}} abstraction
solves both of these problems.
Familiarity with [Pods](/docs/concepts/workloads/pods/) is suggested.

<!-- body -->

## Background

Docker has a concept of
[volumes](https://docs.docker.com/storage/), though it is
somewhat looser and less managed. A Docker volume is a directory on
disk or in another container. Docker provides volume
drivers, but the functionality is somewhat limited.

Kubernetes supports many types of volumes. A {{< glossary_tooltip term_id="pod" text="Pod" >}}

> can use any number of volume types simultaneously. The lifetime of a Kubernetes volume

Reviewer: The note about volume lifetime is not correct. Ephemeral volume types have a lifetime of a pod, but persistent volumes have a lifetime beyond the pod.

Author: Hi @msau42 .
Thanks for the feedback. Initially, I created this PR to clean up the formatting and styling
of the page content. I was also experimenting with adapting some new docs style guidelines.
I went a bit further by modifying some of the text (grammar, readability), but
did not delve too deeply into the content. I'll look again at the page and see if I can make more content changes
otherwise, I can log an issue to review the page in more detail. How does that sound?

Author: Let me see if I can updated correctly.

Author: Do you think it is still relevant to include the paragraph about Docker volumes (as a comparison)?

Reviewer: I think leaving content updates to a future pr is fine. I think the Docker volumes comparison may need to be updated as well. The biggest difference I think that's relevant is Docker volumes write data to the host and do not persist if the pod is restarted on a different node.

coincides with the lifetime of its pod. Consequently, a volume outlives any containers
that run within the pod, and data is preserved across container restarts. When a
pod ceases to exist, the volume is destroyed.

At its core, a volume is just a directory, possibly with some data in it, which
is accessible to the containers in a pod. How that directory comes to be, the
medium that backs it, and the contents of it are determined by the particular
volume type used.

To use a volume, specify the volumes to provide for the Pod in `.spec.volumes`
and declare where to mount those volumes into containers in `.spec.containers[*].volumeMounts`.
A process in a container sees a filesystem view composed from their Docker
image and volumes. The [Docker image](https://docs.docker.com/userguide/dockerimages/)
is at the root of the filesystem hierarchy. Volumes mount at the specified paths within
the image. Volumes can not mount onto other volumes or have hard links to
other volumes. Each Container in the Pod's configuration must independently specify where to
mount each volume.

## Types of Volumes {#volume-types}

> Kubernetes supports several types of volumes.

Reviewer: is there another place we can see the TOC? I think it's convenient to have a condensed list of all the plugins up front and be able to skip to the relevant detailed section depending on what you're interested in

Author: Yes, the TOC for the page (which gets automatically generated from the headings)
is located in the right hand side of the page. The in-page table of contents should contain
each plugin. Is this change okay?

Reviewer: yes thanks for explaining!

### awsElasticBlockStore {#awselasticblockstore}

An `awsElasticBlockStore` volume mounts an Amazon Web Services (AWS)
[EBS volume](https://aws.amazon.com/ebs/) into your pod. Unlike
`emptyDir`, which is erased when a pod is removed, the contents of an EBS

> volume are preserved and the volume is merely unmounted. This means that an

Reviewer: maybe instead of preserved, use persisted, to reflect our persistentvolumes concept

Author: OK. I will change. I did not attempt to change the description for each plugin.

<!-- overview -->

On-disk files in a container are ephemeral, which presents some problems for
non-trivial applications when running in containers. One problem
is the loss of files when a container crashes. The kubelet restarts the container
but with a clean state. A second problem occurs when sharing files
between containers running together in a pod.
The Kubernetes {{< glossary_tooltip text="volume" term_id="volume" >}} abstraction

> solves both of these problems.

Author: This is a good comment, though, I'd rather keep the focus of this PR on style cleanups and minor rewording.

> Familiarity with [Pods](/docs/concepts/workloads/pods/pod/) is suggested.

Author: I will fix.

<!-- body -->

## Background

Docker has a concept of
[volumes](https://docs.docker.com/storage/), though it is
somewhat looser and less managed. A Docker volume is a directory on
disk or in another container. Lifetimes are not managed and until very

> recently there were only local-disk-backed volumes. Docker now provides volume

Author: I decided to leave the original text in a number of sections. I will look at the content again.

drivers, but the functionality is very limited for now (e.g. as of Docker 1.7
only one volume driver is allowed per container and there is no way to pass

> parameters to volumes).

Author: Yes, I wanted to remove this but left in place. Hope someone from Storage would agree to remove or update.

Kubernetes supports many types of volumes. A {{< glossary_tooltip term_id="pod" text="Pod" >}}
can use any number of volume types simultaneously. The lifetime of a Kubernetes volume
coincides with the lifetime of its pod. Consequently, a volume outlives any containers
that run within the pod, and data is preserved across container restarts. When a
pod ceases to exist, the volume is destroyed.

At its core, a volume is just a directory, possibly with some data in it, which
is accessible to the containers in a pod. How that directory comes to be, the
medium that backs it, and the contents of it are determined by the particular
volume type used.

To use a volume, specify the volumes to provide for the Pod in `Pod.spec.volumes`

> and declare where to mount those volumes into containers in `Pod.spec.containers[*].volumeMounts`.

Author: OK. Will update. Went back and forth on this change.

A process in a container sees a filesystem view composed from their Docker
image and volumes. The [Docker image](https://docs.docker.com/userguide/dockerimages/)
is at the root of the filesystem hierarchy. Volumes mount at the specified paths within
the image. Volumes can not mount onto other volumes or have hard links to
other volumes. Each container in the pod must independently specify where to
mount each volume.

## Types of Volumes {#volume-types}

> Kubernetes supports several types of volumes. New contributions are welcome.

Author: OK. I will remove.

### awsElasticBlockStore {#awselasticblockstore}

An `awsElasticBlockStore` volume mounts an Amazon Web Services (AWS)
[EBS volume](https://aws.amazon.com/ebs/) into your pod. Unlike
`emptyDir`, which is erased when a pod is removed, the contents of an EBS
volume are preserved and the volume is merely unmounted. This means that an

> EBS volume can be pre-populated with data, and that data can be transferred between pods.

Author: Changed "handed off" to transferred. Is this an accurate change?

Author: Yes. My goal is to come up with a replacement phrase for "handed off".

Reviewer: shared? or used?

Author: OK. I will change to "shared". This phrase is used throughout the page.

{{< caution >}}
You must create an EBS volume by using `aws ec2 create-volume` or the AWS API before you can use it.
{{< /caution >}}

There are some restrictions when using an `awsElasticBlockStore` volume:

* the nodes on which pods are running must be AWS EC2 instances
* those instances need to be in the same region and availability zone as the EBS volume

> * EBS only supports a single EC2 instance mounting a volume

Author: Will leave this for another PR.

#### Creating an EBS volume

Before you can use an EBS volume with a pod, you need to create it.

```shell
aws ec2 create-volume --availability-zone=eu-west-1a --size=10 --volume-type=gp2
```

Make sure the zone matches the zone you brought up your cluster in. Check that the size and EBS volume
type are suitable for your use.

> #### AWS EBS Pod configuration example

Author: I want a consistent heading for all examples. I can remove "Pod" from the headings.

the pod that encloses it. Consequently, a volume outlives any containers that run
within the pod, and data is preserved across container restarts. Of course, when a
pod ceases to exist, the volume is destroyed. Perhaps more
importantly than this, Kubernetes supports many types of volumes, and a pod can
use any number of them simultaneously.

At its core, a volume is just a directory, possibly with some data in it, which
is accessible to the containers in a pod.  How that directory comes to be, the
medium that backs it, and the contents of it are determined by the particular
volume type used.

To use a volume, a `Pod` specifies what volumes to provide for the `Pod`
by setting the `.spec.volumes` field and where to mount those volumes
into containers (the `.spec.containers[*].volumeMounts` field).
A process in a container sees a filesystem view composed from their Docker
image and volumes. The [Docker image](https://docs.docker.com/userguide/dockerimages/)
is at the root of the filesystem hierarchy. Volumes mount at the specified paths within
the image. Volumes can not mount onto other volumes or have hard links to

> other volumes. Each container in the pod must independently specify where to

Author: Do you think this is referring to the Kubernetes Container resource or containers in general?

> other volumes. Each container in the pod must independently specify where to

Author: I agree that this is sentence is confusing. I've gone over this a few times.
I will update and capitalize as a resource because the sentence is describing how to "specify" the Container and Pod objects.
I also agree that the name of the field is called `container` but if this styling is used I would update the sentence to identify the "field to set".

mount each volume.

## Types of Volumes

Kubernetes supports several types of volumes:

> * [awsElasticBlockStore](#awselasticblockstore)

Author: I removed the in-page Table of contents. The theme provides a TOC which is easier to maintain than a list.
Is this a reasonable change?

> * [awsElasticBlockStore](#awselasticblockstore)

Reviewer 2: Sounds good to me.

  - name: test-volume
    # This AWS EBS volume must already exist.
    awsElasticBlockStore:
      volumeID: "<volume id>"
      fsType: ext4
```

#### AWS EBS CSI migration

{{< feature-state for_k8s_version="v1.17" state="beta" >}}

> The `CSIMigration` feature for `awsElasticBlockStore`, when enabled, shims all plugin operations

Reviewer: do you think "shims" is a clear word to describe this? or would something like "route" or "redirects" be more descriptive?

Author: No, not really. I think the docs could be improved here.

Author: I updated the use of "shim" in the page. PTAL.

from the existing in-tree plugin to the `ebs.csi.aws.com` Container
Storage Interface (CSI) driver. In order to use this feature, the [AWS EBS CSI
driver](https://github.com/kubernetes-sigs/aws-ebs-csi-driver)
must be installed on the cluster and the `CSIMigration` and `CSIMigrationAWS`
beta features must be enabled.

#### AWS EBS CSI migration complete

{{< feature-state for_k8s_version="v1.17" state="alpha" >}}

To disable the `awsElasticBlockStore` storage plugin from being loaded by the controller manager
and the kubelet, set the `CSIMigrationAWSComplete` flag to `true`. This feature requires the `ebs.csi.aws.com` Container Storage Interface (CSI) driver installed on all worker nodes.

> ### azureDisk {#azuredisk}

Reviewer: the doc seems to be inconsistent about which volume types are persistent and which are ephemeral. Would it be more clear if we broke up this list of volume types into two sections, ephemeral and persistent?

Author: I agree.

  - name: test-volume
    # This AWS EBS volume must already exist.
    awsElasticBlockStore:
      volumeID: "<volume id>"
      fsType: ext4
```

#### AWS CSI migration

{{< feature-state for_k8s_version="v1.17" state="beta" >}}

The `CSIMigration` feature for `awsElasticBlockStore`, when enabled, shims all plugin operations
from the existing in-tree plugin to the `ebs.csi.aws.com` Container
Storage Interface (CSI) driver. In order to use this feature, the [AWS EBS CSI
driver](https://github.com/kubernetes-sigs/aws-ebs-csi-driver)
must be installed on the cluster and the `CSIMigration` and `CSIMigrationAWS`
Beta features must be enabled.

> #### AWS CSI migration complete

Author: This is a common pattern in the page.
This is about using the feature gate, `CSIMigrationAWSComplete` to disable the plugin.
As for when this is used, it is hard to tell.
/cc @kubernetes/sig-storage-pr-reviews

> #### AWS CSI migration complete

Author: I'd rather not use a level 5 heading.

> #### AWS CSI migration complete

Reviewer 2: I suppose another option is to move (separate PR?) the CSI migration task descriptions into a separate page, perhaps in the Tasks section?

What do you think @kbhawkey ? If that's future work we'd plan to pick up, that could be hint to leave things as-is here _for now_.

  - name: test-volume
    # This AWS EBS volume must already exist.
    awsElasticBlockStore:
      volumeID: "<volume id>"
      fsType: ext4
```

> #### CSIMigrationAWS

Author: Yes, I agree with you. I need to look at the the text that follows and make sure the feature gate name
is referenced (feature gate page). There should be unique headings for each section.

> #### CSIMigrationAWS

Author: I changed the headings. Let me know what you think. Thanks.

> #### CSIMigrationAWS

Author: How are the headings now?

See the [CephFS example](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/volumes/cephfs/) for more details.

### cinder

{{< note >}}
Kubernetes must be configured with the OpenStack Cloud Provider.
{{< /note >}}

> The `cinder` volume plugin is used to mount the OpenStack Cinder volume into your pod.

Author: Is this original sentence technically incorrect?

> The `cinder` volume plugin is used to mount the OpenStack Cinder volume into your pod.

Reviewer 2: The original sentence looks technically correct, I think the grammar is a bit off (“the OpenStack Cinder volume”, singular, doesn't feel quite right). Just a nit though.

  - name: test-volume
    # This OpenStack volume must already exist.
    cinder:
      volumeID: "<volume id>"
      fsType: ext4
```

#### OpenStack CSI migration

{{< feature-state for_k8s_version="v1.18" state="beta" >}}

The `CSIMigration` feature for Cinder, when enabled, shims all plugin operations
from the existing in-tree plugin to the `cinder.csi.openstack.org` Container
Storage Interface (CSI) Driver. In order to use this feature, the [Openstack Cinder CSI
Driver](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-cinder-csi-plugin.md)
must be installed on the cluster and the `CSIMigration` and `CSIMigrationOpenStack`
Beta features must be enabled.

### configMap

> A [`ConfigMap`](/docs/tasks/configure-pod-container/configure-pod-configmap/)

Author: Yes, I think I left this as a test. Will update.

### cinder {#cinder}

{{< note >}}

> Kubernetes must be configured with the OpenStack Cloud Provider.

Author: OK will fix.

{{< /note >}}

`cinder` is used to mount OpenStack Cinder volume into your pod.

> #### cinder volume example configuration

Reviewer 2: Think this is resolved.

  - name: test-volume
    # This OpenStack volume must already exist.
    cinder:
      volumeID: "<volume id>"
      fsType: ext4
```

#### CSIMigrationOpenStack

{{< feature-state for_k8s_version="v1.18" state="beta" >}}

The CSIMigration feature for Cinder, when enabled, shims all plugin operations
from the existing in-tree plugin to the `cinder.csi.openstack.org` Container
Storage Interface (CSI) Driver. In order to use this feature, the [Openstack Cinder CSI
Driver](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/using-cinder-csi-plugin.md)
must be installed on the cluster and the `CSIMigration` and `CSIMigrationOpenStack`
Beta features must be enabled.

### ConfigMap {#configmap}

[`ConfigMap`](/docs/tasks/configure-pod-container/configure-pod-configmap/)
provides a way to inject configuration data into pods.
The data stored in a `ConfigMap` can be referenced in a volume of type
`configMap` and then consumed by containerized applications running in a pod.

> When referencing a ConfigMap, you provide the name of the ConfigMap in the

Author: Good questions.

receive ConfigMap updates.
{{< /note >}}

{{< note >}}
Text data is exposed as files using the UTF-8 character encoding. For other character encodings, use binaryData.
{{< /note >}}

### downwardAPI {#downwardapi}

A `downwardAPI` volume makes downward API data available to applications.
It mounts a directory and writes the requested data in plain text files.

{{< note >}}
A container using downward API as a [`subPath`](#using-subpath) volume mount will not
receive downward API updates.
{{< /note >}}

See the [`downwardAPI` volume example](/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/) for more details.

### emptyDir {#emptydir}

An `emptyDir` volume is first created when a pod is assigned to a node, and
exists as long as that Pod is running on that node. As the name says, the
`emptyDir` volume is initially empty. All containers in the Pod can read and write the same
files in the `emptyDir` volume, though that volume can be mounted at the same
or different paths in each container. When a pod is removed from a node for
any reason, the data in the `emptyDir` is deleted permanently.

{{< note >}}

> A container crashing does *not* remove a pod from a node, so the data in an `emptyDir` volume is safe across container crashes.

Author: Yes. When there is a description about a resource with the following actions:
creating, running, deleting, ...
I think it makes sense to identify the resource using its type or `kind`.
Will fix.

{{< /note >}}

Some uses for an `emptyDir` are:

* scratch space, such as for a disk-based merge sort
* checkpointing a long computation for recovery from crashes
* holding files that a content-manager container fetches while a webserver
  container serves the data

By default, `emptyDir` volumes are stored on whatever medium is backing the
node such as disk or SSD, or network storage, depending on your
environment. However, you can set the `emptyDir.medium` field to `"Memory"`
to tell Kubernetes to mount a tmpfs (RAM-backed filesystem) for you instead.
While tmpfs is very fast, be aware that unlike disks, tmpfs is cleared on
node reboot and any files you write will count against your container's

> memory limit.

Author: Thanks for the comment. Is this important to add here?

{{< /caution >}}

{{< note >}}
A container using a ConfigMap as a [`subPath`](#using-subpath) volume mount will not
receive ConfigMap updates.
{{< /note >}}

{{< note >}}
Text data is exposed as files using the UTF-8 character encoding. To use some other character encoding, use binaryData.
{{< /note >}}

### downwardAPI {#downwardapi}

A `downwardAPI` volume makes downward API data available to applications.
It mounts a directory and writes the requested data in plain text files.

{{< note >}}
A container using Downward API as a [`subPath`](#using-subpath) volume mount will not

> receive Downward API updates.

Author: Yes, this needs more cleanup.

> receive Downward API updates.

Author: I changed this. Not sure if correct.
downwardAPI is a field and `DownwardAPIVolumeSource` is a resource.

`emptyDir` volume is initially empty. All containers in the Pod can read and write the same
files in the `emptyDir` volume, though that volume can be mounted at the same
or different paths in each container. When a Pod is removed from a node for
any reason, the data in the `emptyDir` is deleted forever.

{{< note >}}
A Container crashing does *NOT* remove a Pod from a node, so the data in an `emptyDir` volume is safe across container crashes.
{{< /note >}}

Some uses for an `emptyDir` are:

* scratch space, such as for a disk-based merge sort
* checkpointing a long computation for recovery from crashes
* holding files that a content-manager container fetches while a webserver
  container serves the data

By default, `emptyDir` volumes are stored on whatever medium is backing the
node such as disk or SSD or network storage, depending on your
environment. However, you can set the `emptyDir.medium` field to `"Memory"`
to tell Kubernetes to mount a tmpfs (RAM-backed filesystem) for you instead.
While tmpfs is very fast, be aware that unlike disks, tmpfs is cleared on
node reboot and any files you write will count against your container's
memory limit.

> #### emptyDir example Pod

Author: I'd like to standardize on this heading. Not sure if a heading is needed.

> #### emptyDir example Pod

Reviewer 2: +0 (soft +1) on removing those headings.

> #### emptyDir example Pod

Author: heading remain

### fc (fibre channel) {#fc}

An `fc` volume type allows an existing fibre channel block storage volume to be mounted in a Pod.
You can specify single or multiple target world wide names (WWNs) using the parameter
`targetWWNs` in your volume configuration. If multiple WWNs are specified,
targetWWNs expect that those WWNs are from multi-path connections.

{{< note >}}
You must configure FC SAN Zoning to allocate and mask those LUNs (volumes) to the target WWNs beforehand so that Kubernetes hosts can access them.
{{< /note >}}

See the [fibre channel example](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/staging/volumes/fibre_channel) for more details.

> ### flocker

Reviewer: we should probably mark this as deprecated/unsupported

Author: OK. I will update. When should a deprecated volume type be removed from the documentation?

`targetWWNs` in your volume configuration. If multiple WWNs are specified,
targetWWNs expect that those WWNs are from multi-path connections.

{{< caution >}}
You must configure FC SAN Zoning to allocate and mask those LUNs (volumes) to the target WWNs beforehand so that Kubernetes hosts can access them.
{{< /caution >}}

See the [fibre channel example](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/staging/volumes/fibre_channel) for more details.

### flocker

[Flocker](https://github.com/ClusterHQ/flocker) is an open-source, clustered
container data volume manager. Flocker provides management
and orchestration of data volumes backed by a variety of storage backends.

A `flocker` volume allows a Flocker dataset to be mounted into a pod. If the
dataset does not already exist in Flocker, it needs to be first created with the Flocker
CLI or by using the Flocker API. If the dataset already exists it will be
reattached by Flocker to the node that the pod is scheduled. This means data
can be transferred between pods as required.

{{< caution >}}
You must have your own Flocker installation running before you can use it.

> {{< /caution >}}

Author: Sure, makes sense.

See the [Flocker example](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/staging/volumes/flocker) for more details.

### gcePersistentDisk

A `gcePersistentDisk` volume mounts a Google Compute Engine (GCE)

> [persistent disk](https://cloud.google.com/compute/docs/disks) (PD) into your pod.

Author: Looking at the GCE docs, there are plenty of instances of "persistent disk" not "Persistent Disk".
Which is correct?

There are some restrictions when using a `gcePersistentDisk`:

* the nodes on which Pods are running must be GCE VMs
* those VMs need to be in the same GCE project and zone as the persistent disk

One feature of GCE persistent disk is concurrent read-only access to a persistent disk.
A `gcePersistentDisk` volume permits multiple consumers to simultaneously
mount a persistent disk as read-only. This means that you can pre-populate a PD with your dataset
and then serve it in parallel from as many Pods as you need. Unfortunately,
PDs can only be mounted by a single consumer in read-write mode. Simultaneous
writers are not allowed.

> Using a persistent disk on a Pod controlled by a ReplicationController will fail unless

Author: May change back to "PD".
Using a PD in a Pod
Yes, can update to ReplicaSet

the PD is read-only or the replica count is 0 or 1.

#### Creating a persistent disk

Before you can use a GCE persistent disk with a Pod, you need to create it.

```shell
gcloud compute disks create --size=500GB --zone=us-central1-a my-data-disk
```

> #### Persistent disk Pod configuration example

Author: mulling this one over

### gitRepo (deprecated) {#gitrepo}

{{< warning >}}

> The `gitRepo` volume type is deprecated. To provision a container with a git repo, mount an [EmptyDir](#emptydir) into an InitContainer that clones the repo using git, then mount the [EmptyDir](#emptydir) into the Pod's container.

Author: I left this section alone as the feature is deprecated. I was tempted to remove it.

      type: FileOrCreate
```

### iscsi

An `iscsi` volume allows an existing iSCSI (SCSI over IP) volume to be mounted
into your Pod. Unlike `emptyDir`, which is erased when a Pod is removed, the
contents of an `iscsi` volume are preserved and the volume is merely
unmounted. This means that an iscsi volume can be pre-populated with data, and
that data can be transferred between pods.

{{< caution >}}
You must have your own iSCSI server running with the volume created before you can use it.

> {{< /caution >}}

Author: I am trying to be consistent. There are cautions and notes in use throughout the page.

A feature of iSCSI is that it can be mounted as read-only by multiple consumers
simultaneously. This means that you can pre-populate a volume with your dataset
and then serve it in parallel from as many Pods as you need. Unfortunately,
iSCSI volumes can only be mounted by a single consumer in read-write mode.
Simultaneous writers are not allowed.

See the [iSCSI example](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/volumes/iscsi) for more details.

### local

A `local` volume represents a mounted local storage device such as a disk,
partition or directory.

Local volumes can only be used as a statically created PersistentVolume. Dynamic
provisioning is not supported.

Compared to `hostPath` volumes, local volumes can be used in a durable and
portable manner without manually scheduling pods to nodes, as the system is aware
of the volume's node constraints by looking at the node affinity on the PersistentVolume.

However, local volumes are still subject to the availability of the underlying
node and are not suitable for all applications. If a node becomes unhealthy,

> then the local volume will also become inaccessible, and a pod using it will not

Author: Sorting this out.

lifecycle.
{{< /note >}}

### nfs

An `nfs` volume allows an existing NFS (Network File System) share to be

> mounted into your pod. Unlike `emptyDir`, which is erased when a pod is

Author: mounted into a pod vs. Pod is removed? OR
mounted into a Pod ... Pod is removed?

See the information about [PersistentVolumes](/docs/concepts/storage/persistent-volumes/) for more
details.

### portworxVolume {#portworxvolume}

A `portworxVolume` is an elastic block storage layer that runs hyperconverged with
Kubernetes. [Portworx](https://portworx.com/use-case/kubernetes-storage/) fingerprints storage
in a server, tiers based on capabilities, and aggregates capacity across multiple servers.
Portworx runs in-guest in virtual machines or on bare metal Linux nodes.

A `portworxVolume` can be dynamically created through Kubernetes or it can also
be pre-provisioned and referenced inside a Pod.
Here is an example Pod referencing a pre-provisioned Portworx volume:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-portworx-volume-pod
spec:
  containers:
  - image: k8s.gcr.io/test-webserver
    name: test-container
    volumeMounts:
    - mountPath: /mnt
      name: pxvol
  volumes:
  - name: pxvol
    # This Portworx volume must already exist.
    portworxVolume:
      volumeID: "pxvol"
      fsType: "<fs-type>"
```

{{< caution >}}
Make sure you have an existing PortworxVolume with name `pxvol`
before using it in the Pod.

> {{< /caution >}}

Author: thinking on this ... consistency throughout the page ...

1.11 and promoted to Beta in 1.12.
To enable this feature on 1.11, you need to explicitly set the `TokenRequestProjection`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/) to

> `true`.

Author: I'll remove it

CSI is the recommended plugin to use Quobyte volumes inside Kubernetes. Quobyte's
GitHub project has [instructions](https://github.com/quobyte/quobyte-csi#quobyte-csi) for deploying Quobyte using CSI, along with examples.

### rbd

An `rbd` volume allows a
[Rados Block Device](https://ceph.com/docs/master/rbd/rbd/) (RBD) volume to be mounted into your
Pod. Unlike `emptyDir`, which is erased when a pod is removed, the contents of
a `rbd` volume are preserved and the volume is merely unmounted. This
means that a RBD volume can be pre-populated with data, and that data can
be transferred between pods.

{{< note >}}
You must have a Ceph installation running before you can use RBD.
{{< /note >}}

A feature of RBD is that it can be mounted as read-only by multiple consumers
simultaneously. This means that you can pre-populate a volume with your dataset
and then serve it in parallel from as many pods as you need. Unfortunately,
RBD volumes can only be mounted by a single consumer in read-write mode.
Simultaneous writers are not allowed.

See the [RBD example](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/volumes/rbd)
for more details.

> ### scaleIO {#scaleio}

Reviewer: this should be marked as deprecated

Author: OK

a `rbd` volume are preserved and the volume is merely unmounted. This
means that a RBD volume can be pre-populated with data, and that data can
be transferred between pods.

{{< caution >}}
You must have your own Ceph installation running before you can use RBD.
{{< /caution >}}

A feature of RBD is that it can be mounted as read-only by multiple consumers
simultaneously. This means that you can pre-populate a volume with your dataset
and then serve it in parallel from as many pods as you need. Unfortunately,
RBD volumes can only be mounted by a single consumer in read-write mode - no
simultaneous writers allowed.

See the [RBD example](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/volumes/rbd)
for more details.

### scaleIO {#scaleio}

ScaleIO is a software-based storage platform that uses existing hardware to
create clusters of scalable shared block networked storage. The `scaleIO` volume
plugin allows deployed pods to access existing ScaleIO
volumes. For information about dynamically provisioning new volumes for persistent volume claims, see
[ScaleIO Persistent Volumes](/docs/concepts/storage/persistent-volumes/#scaleio).

> {{< caution >}}

Author: Will look at the set of callouts and decide.

      fsType: xfs
```

For further details, see the [ScaleIO](https://github.com/kubernetes/examples/tree/{{< param "githubbranch" >}}/staging/volumes/scaleio) examples.

### secret

A `secret` volume is used to pass sensitive information, such as passwords, to
pods. You can store secrets in the Kubernetes API and mount them as files for
use by pods without coupling to Kubernetes directly. `secret` volumes are
backed by tmpfs (a RAM-backed filesystem) so they are never written to
non-volatile storage.

{{< caution >}}
You must create a Secret in the Kubernetes API before you can use it.
{{< /caution >}}

{{< note >}}
A container using a Secret as a [`subPath`](#using-subpath) volume mount will not
receive Secret updates.
{{< /note >}}

> For more details, see [Configuring Secrets](/docs/concepts/configuration/secret/).

Author: Yes, this should be linking to that page

> For more details, see [Configuring Secrets](/docs/concepts/configuration/secret/).

Author: ?

* `objectspacereservation`
* `iopslimit`

Existing volumes created using these parameters will be migrated to the vSphere CSI driver,
but new volumes created by the vSphere CSI driver will not be honoring these parameters.
{{< /note >}}

#### vSphere CSI migration complete {#vsphere-csi-migration-complete}

{{< feature-state for_k8s_version="v1.19" state="beta" >}}

To turn off the `vsphereVolume` plugin from being loaded by the controller manager and the kubelet, you need to set this feature flag to `true`. You must install a `csi.vsphere.vmware.com` {{< glossary_tooltip text="CSI" term_id="csi" >}} driver on all worker nodes.

## Using subPath {#using-subpath}

Sometimes, it is useful to share one volume for multiple uses in a single pod. The `volumeMounts.subPath`
property can be used to specify a sub-path inside the referenced volume instead of its root.

> Here is an example of a `Pod` with a LAMP stack (Linux Apache Mysql PHP) using a single, shared volume.

Author: Yes, testing the styling.

The HTML contents are mapped to its `html` folder and the databases are stored in its `mysql` folder:

```yaml

> apiVersion: v1

Reviewer 2: aside: we ought to add a comment that this Pod is an example of using `subPath`, and isn't meant to indicate recommended practise.

Author: OK. Thanks.

        claimName: my-lamp-site-data
```

### Using subPath with expanded environment variables {#using-subpath-expanded-environment}

{{< feature-state for_k8s_version="v1.17" state="stable" >}}

Use the `subPathExpr` field to construct `subPath` directory names from downward API environment variables.
The `subPath` and `subPathExpr` properties are mutually exclusive.

In this example, a `Pod` uses `subPathExpr` to create a directory `pod1` within the `hostPath` volume `/var/log/pods`.

> The `hostPath` volume takes the `Pod` name from the `downwardAPI`. The host directory `/var/log/pods/pod1` is mounted at `/logs` in the container.

Reviewer 2: :+1: Nice rewording.

## Resources

The storage media (such as Disk or SSD) of an `emptyDir` volume is determined by the
medium of the filesystem holding the kubelet root dir (typically
`/var/lib/kubelet`). There is no limit on how much space an `emptyDir` or
`hostPath` volume can consume, and no isolation between containers or between

> pods.

Reviewer 2: Aside: this is slightly out of date / inaccurate.
1. You can have memory-backed `emptyDir` volumes.
1. Calling `/var/lib/kubelet` a root directory of any kind is a bit misleading.

Reviewer 2: Earlier, the page mentions that you can set the `emptyDir.medium` field to `"Memory"` to get memory-backed ephemeral local storage.

Author: Yes, I read through this and found that the content is confusing (or out of place).

> In the future, we expect that `emptyDir` and `hostPath` volumes will be able to

Reviewer 2: Aside: this is partly done.

Author: Pending

request a certain amount of space using a [resource](/docs/concepts/configuration/manage-resources-containers/)
specification, and to select the type of media to use, for clusters that have
several media types.

## Out-of-tree volume plugins

> The out-of-tree volume plugins include the `Container Storage Interface (CSI)`

Reviewer 2: ```suggestion
The out-of-tree volume plugins include the {{< glossary_tooltip text="Container Storage Interface" term_id="csi" >}} (CSI)
```
?

Author: I'll consider a tooltip. Let me look if CSI has been referenced prior to this section.

> and `FlexVolume`. They enable storage vendors to create custom storage plugins

Reviewer 2: I'd leave the original. FlexVolume is the technology; `flexVolume` the volume type.

Author: OK

without adding them to the Kubernetes repository.

Previously, all volume plugins were "in-tree". The "in-tree" plugins were built, linked, compiled,
and shipped with the core Kubernetes binaries. This meant that adding a new storage system to
Kubernetes (a volume plugin) required checking code into the core Kubernetes code repository.

> Both `CSI` and `FlexVolume` allow volume plugins to be developed independent of

Reviewer 2: ```suggestion
The `csi` and `flexVolume` volume types both allow storage plugins to be developed independent of
```
?

Author: Previously, csi and flexvolume are called volume plugins (line 1251).

the Kubernetes code base, and deployed (installed) on Kubernetes clusters as
extensions.

For storage vendors looking to create an out-of-tree volume plugin, please refer
to the [volume plugin FAQ](https://github.com/kubernetes/community/blob/master/sig-storage/volume-plugin-faq.md).

> ### CSI

Reviewer 2: ```suggestion
### csi
```
(to match the other headings, that are capitalized the same as the volume type)

the Kubernetes code base, and deployed (installed) on Kubernetes clusters as
extensions.

For storage vendors looking to create an out-of-tree volume plugin, please refer
to [this FAQ](https://github.com/kubernetes/community/blob/master/sig-storage/volume-plugin-faq.md).
to the [volume plugin FAQ](https://github.com/kubernetes/community/blob/master/sig-storage/volume-plugin-faq.md).

### CSI

[Container Storage Interface](https://github.com/container-storage-interface/spec/blob/master/spec.md) (CSI)
defines a standard interface for container orchestration systems (like
[Container Storage Interface](https://github.com/container-storage-interface/spec/blob/master/spec.md)
(CSI) defines a standard interface for container orchestration systems (like
Kubernetes) to expose arbitrary storage systems to their container workloads.

Please read the [CSI design proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md) for more information.

CSI support was introduced as alpha in Kubernetes v1.9, moved to beta in
Kubernetes v1.10, and is GA in Kubernetes v1.13.

> 

Reviewer 2: :+1:

{{< /note >}}

Once a CSI compatible volume driver is deployed on a Kubernetes cluster, users
may use the `csi` volume type to attach or mount the volumes exposed by the
CSI driver.

A `csi` volume can be used in a `Pod` in three different ways:

* through a reference to a [PersistentVolumeClaim](#persistentvolumeclaim)

> * with a [generic ephemeral volume](/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volume) (Alpha feature)

Reviewer 2: I wouldn't capitalize Alpha. It's an adjective, the way I conceive it, so just as you wouldn't capitalize “new”, “deprecated” or “upcoming” I wouldn't capitalize this one.

Author: Yes, I guess so. I'd rather remove it.

persist across pod restarts. See [Ephemeral
Volumes](/docs/concepts/storage/ephemeral-volumes/#csi-ephemeral-volume)
for more information.

For more information on how to develop a CSI driver, refer to the
[kubernetes-csi documentation](https://kubernetes-csi.github.io/docs/)

#### Migrating to CSI drivers from in-tree plugins

{{< feature-state for_k8s_version="v1.14" state="alpha" >}}

The `CSIMigration` feature, when enabled, directs operations against existing in-tree
plugins to corresponding CSI plugins (which are expected to be installed and configured).
The feature implements the necessary translation logic and shims to re-route the
operations in a seamless fashion. As a result, operators do not have to make any
configuration changes to existing Storage Classes, PVs or PVCs (referring to
in-tree plugins) when transitioning to a CSI driver that supersedes an in-tree plugin.

The operations and features that are supported include:
provisioning/delete, attach/detach, mount/unmount and resizing of volumes.

In-tree plugins that support `CSIMigration` and have a corresponding CSI driver implemented
are listed in [Types of Volumes](#volume-types).

> ### FlexVolume {#flexVolume}

Reviewer 2: ```suggestion
### flexVolume
```
(to match the other headings, that are capitalized the same as the volume type)

> `FlexVolume` is an out-of-tree plugin interface that has existed in Kubernetes

Reviewer 2: The original is correct: the _plugin interface_ is FlexVolume.

since version 1.2 (before CSI). It uses an exec-based model to interface with

> drivers. The `FlexVolume` driver binaries must be installed in a pre-defined volume

Reviewer 2: ```suggestion
drivers. The FlexVolume driver binaries must be installed in a pre-defined volume
```

Author: Not sure about this one. I consider a driver name similar to a namespace name, or plugin name, or library name.

> plugin path on each node and in some cases the control plane node.

Reviewer 2: ```suggestion
plugin path on each node and in some cases the control plane node(s) as well.
```
?

Author: Is it in addition to?

Reviewer 2: The point here is that “master” can be a plural noun but “control plane node” is always singular. Clusters have run a replicated control plane.

Author: Right. I agree that "control plane node" is singular. Sure, can make node plural. Thanks!

> Pods interact with `FlexVolume` drivers through the `flexvolume` in-tree plugin.

Reviewer 2: ```suggestion
Pods interact with FlexVolume drivers through the `flexVolume` in-tree volume type.
```

## Mount propagation

Mount propagation allows for sharing volumes mounted by a container to
other containers in the same pod, or even to other pods on the same node.

Mount propagation of a volume is controlled by the `mountPropagation` field in `Container.volumeMounts`.
Its values are:

* `None` - This volume mount will not receive any subsequent mounts
  that are mounted to this volume or any of its subdirectories by the host.
  In similar fashion, no mounts created by the container will be visible on
  the host. This is the default mode.

  This mode is equal to `private` mount propagation as described in the
  [Linux kernel documentation](https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt)

* `HostToContainer` - This volume mount will receive all subsequent mounts
  that are mounted to this volume or any of its subdirectories.

  In other words, if the host mounts anything inside the volume mount, the
  container will see it mounted there.

  Similarly, if any Pod with `Bidirectional` mount propagation to the same

>   volume mounts anything there, the `Container` with `HostToContainer` mount

Reviewer 2: I'd write:
```suggestion
  volume mounts anything there, the container with `HostToContainer` mount
```
or: `container`

  propagation will see it.

  This mode is equal to `rslave` mount propagation as described in the
  [Linux kernel documentation](https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt)

* `Bidirectional` - This volume mount behaves the same the `HostToContainer` mount.
  In addition, all volume mounts created by the container will be propagated
  back to the host and to all containers of all pods that use the same volume.

>   A typical use case for this mode is a Pod with a `FlexVolume` or `CSI` driver or

Reviewer 2: I'd put:
```suggestion
  A typical use case for this mode is a Pod using a `flexVolume` or `csi` volume, or
```

  a Pod that needs to mount something on the host using a `hostPath` volume.

  This mode is equal to `rshared` mount propagation as described in the
  [Linux kernel documentation](https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt)

  {{< caution >}}
  `Bidirectional` mount propagation can be dangerous. It can damage
  the host operating system and therefore it is allowed only in privileged
  containers. Familiarity with Linux kernel behavior is strongly recommended.
  In addition, any volume mounts created by containers in pods must be destroyed
  (unmounted) by the containers on termination.

>   {{< /caution >}}

Reviewer 2: This might merit a `{{< warning >}}`. (OK to save that tweak for a later PR).

{{< feature-state for_k8s_version="v1.18" state="stable" >}}

Vendors with external CSI drivers can implement raw block volume support
in Kubernetes workloads.

You can set up your
[PersistentVolume/PersistentVolumeClaim with raw block volume support](/docs/concepts/storage/persistent-volumes/#raw-block-volume-support) as usual, without any CSI specific changes.

#### CSI ephemeral volumes

{{< feature-state for_k8s_version="v1.16" state="beta" >}}

You can directly configure CSI volumes within the Pod
specification. Volumes specified in this way are ephemeral and do not
persist across pod restarts. See [Ephemeral
Volumes](/docs/concepts/storage/ephemeral-volumes/#csi-ephemeral-volume)
for more information.

For more information on how to develop a CSI driver, refer to the
[kubernetes-csi documentation](https://kubernetes-csi.github.io/docs/)

#### Migrating to CSI drivers from in-tree plugins

> {{< feature-state for_k8s_version="v1.14" state="alpha" >}}

Reviewer: this has been beta since 1.17

Author: OK. I will update.
